### ISSUE TITLE
MCprep Internal Refactoring

### ISSUE THREAD

**Issue by StandingPadAnimations** [TheDuckCow/MCprep#346]
The MCprep codebase is at least 9 years old now, which means some things definitely need to be refactored. This list will keep track of everything that needs refactoring. These include:
* The entirety of `conf.py`
* Meshswap and other functions that are massive
* Util functions.

**Comment by TheDuckCow** [comment]
Hey @StandingPadAnimations, thanks for setting this up. I expanded on your description (and we'll continue to do so) to include more details. Let's be explicit about _what_ refactoring is being done, just to avoid steps that are sideways. Want to make sure we make changes that are objective improvements. 

What are you thinking about changing with the util functions (and did you have anything different in mind for conf.py above)?

**Comment by StandingPadAnimations** [comment]
> What are you thinking about changing with the util functions (and did you have anything different in mind for conf.py above)?

Util functions might get minor changes (depends on if they need to be changed or not). As for `conf.py`, I think I should rephrase it to "change most of the global variables"

The way global variables are used in `conf.py` would cause issues with developers trying to use variables with the same names (if they didn't know they were already used). Also, in many other languages (like C++ and Rust), global variables are considered bad practice (for a multitude of reasons, such as global variables from static or dynamic libraries usable on some OSes, but not others, and the extra runtime penalty for compiled languages in order to create those global variables). The solution I have in mind is a singleton class (basically a class with a single instance that is reused when you instantiate it) that contains the variables.

Of course that might cause issues, so the original variables won't be removed and instead be deprecated (the plan is to remove them in the release after next release). The same applies to any changes to integral parts of MCprep. To track these changes, I'll make a table that contains all deprecated things with their replacements and an example on using those replacements.

**Comment by TheDuckCow** [comment]
Agreed, globals are not ideal and I agree we should move away from them. Though just to clarify this point, the globals only exist within the scope of that one file. 

Conf I think can still be a `conf.py` file (renaming python files always leads to module update errors, unfortunately). I would suggest turning the majority of conf.py into a singleton class. The structure would be something like this:

```python

class AddonEnv:
    dev = False 
    
    # Used to print out extra information, set false with distribution
    verbose = True
    
    def __init__(self):
        # ... do stuff to check for env.ini or whatever startup stuff.

    def log(self, *args):
        if self.verbose:
            print(*args)

    def debug(self, *args)
        if self.dev and self.verbose:
            print(*args)
...

env = AddonEnv()

```

Then in other files, we would do something like:

```python
from .conf import env

env.debug("This is a debug print statement!")

if env.dev:
    print("Dev env active, do things differently")

```

In terms of not removing them right away, yes.. could be good to have the basics of the new class in place without being used for one release, to avoid the updating issues, and then in the next release after do the full migration. Let me know what you think of this @StandingPadAnimations!

**Comment by StandingPadAnimations** [comment]
With a singleton class, there would be no need for a global instance. Code would be written as:
```py
def register():
	env = AddonEnv() # Create an instance and set values
	...

# Somewhere else
def foo()
	env = AddonEnv() # Same instance as the one created in register
```

As for full migrations, that would be handled with `internal-rewrites` anyway, so most developers can focus on developing new features (provided they don't accidentally use removed features, but that should be easy for maintainers to check). The idea is to make internal refactoring as painless as possible.

I think it would also be nice to use this opportunity to write documentation on internal MCprep features (such as the functions in `util.py`), but that's a separate issue.

**Comment by StandingPadAnimations** [comment]
Closing as done
